service/history/shard/ownership_scaled_rate_burst.go

```diff
@@ -0,0 +1,151 @@
+)
+
+// Rate returns the rate of the base rate limiter multiplied by the shard ownership share.
+func (rb *OwnershipScaledRateBurst) Rate() float64 {
+	return rb.rb.Rate() * rb.scaleFactor()
+}
+
+// Burst returns the burst quota of the base rate limiter multiplied by the shard ownership share, rounded up to the
+// nearest integer. We round up because we don't want to let this drop to zero unless the base burst is zero.
+func (rb *OwnershipScaledRateBurst) Burst() int {
+	return int(math.Ceil(float64(rb.rb.Burst()) * rb.scaleFactor()))
+}
+
+// scaleFactor returns the fraction of the total shards in the cluster owned by this host. It returns 1.0 if there
+// haven't been any shard count updates yet.
+func (rb *OwnershipScaledRateBurst) scaleFactor() float64 {
+	shardCount := rb.shardCount.Load()
+	if shardCount == shardCountNotSet {
+		// If the shard count is not set, then we haven't received the first shard count update yet. In this case, we
+		// return 1.0 so that the base rate/burst quotas are not scaled.
+		return 1.0
+	}
+
+	return float64(shardCount) / float64(rb.totalShards)
+}
+
+// NewOwnershipScaledRateBurst returns a OwnershipScaledRateBurst that scales the rate/burst quotas of the base
+// RateBurst by the fraction of the total shards in the cluster owned by this host. The updateAppliedCallback field is a
+// channel which is sent to in a blocking fashion when the shard count updates are applied. This is useful for testing.
+// In production, you should pass in nil, which will cause the callback to be ignored. If totalNumShards is
+// non-positive, then an error is returned.
+func NewOwnershipScaledRateBurst(
+	rb quotas.RateBurst,
+	shardCounter ShardCounter,
+	totalNumShards int,
+	updateAppliedCallback chan struct{},
+) (*OwnershipScaledRateBurst, error) {
+	if totalNumShards <= 0 {
+		return nil, fmt.Errorf("%w: %d", ErrNonPositiveTotalNumShards, totalNumShards)
+	}
+	subscription := shardCounter.Subscribe()
+	srb := &OwnershipScaledRateBurst{
+		rb:                    rb,
+		totalShards:           totalNumShards,
+		subscription:          subscription,
+		updateAppliedCallback: updateAppliedCallback,
+	}
+	// Initialize the shard count to the shardCountNotSet sentinel value so that we don't try to apply the scale factor
+	// until we receive the first shard count.
+	srb.shardCount.Store(shardCountNotSet)
+	srb.wg.Add(1)
+
+	go srb.startScaling()
+
+	return srb, nil
+}
+
+func (rb *OwnershipScaledRateBurst) startScaling() {
+	defer rb.wg.Done()
+
+	for shardCount := range rb.subscription.ShardCount() {
+		rb.shardCount.Store(int64(shardCount))
+		if rb.updateAppliedCallback != nil {
+			rb.updateAppliedCallback <- struct{}{}
+		}
+	}
+}
+
+// StopScaling unsubscribes from the shard counter and stops scaling the rate and burst quotas. This method blocks until
+// the shard count subscription goroutine exits (which should be almost immediately).
+func (rb *OwnershipScaledRateBurst) StopScaling() {
```

Comment:
who will call this method? fx shutdown logic?

Reply:
Yeah, exactly. This is a shutdown hook.